stats/patient_interceptor.go

```diff
@@ -0,0 +1,108 @@
+package stats
+
+import (
+	"errors"
+	"log"
+
+	"github.com/intervention-engine/fhir/models"
+)
+
+// PatientStatsCreateInterceptor intercepts any new patient resources added to the database
+// and updates the Synthetic Mass population statistics based on that patient's address.
+type PatientStatsCreateInterceptor struct {
+	DataAccess StatsDataAccess
+}
+
+// Before is unused
+func (s *PatientStatsCreateInterceptor) Before(resource interface{}) {}
+
+// After increments population statistics after a patient resource is created.
+func (s *PatientStatsCreateInterceptor) After(resource interface{}) {
+	patient, ok := resource.(*models.Patient)
+
+	if ok {
+		err := s.DataAccess.AddPatientStat(patient)
+		if err != nil {
+			log.Printf("PatientStatsCreateInterceptor: After: %s\n", err.Error())
+		}
+	}
+}
+
+// OnError is unused
+func (s *PatientStatsCreateInterceptor) OnError(err error, resource interface{}) {}
+
+// PatientStatsUpdateInterceptor intercepts any updated patient resources
+// and updates the Synthetic Mass population statistics based on that patient's address.
+type PatientStatsUpdateInterceptor struct {
+	DataAccess StatsDataAccess
+	// The state of the patient before the database update, for comparison after the database update
+	patientBefore *models.Patient
+	// Tracks if the interceptor failed to cache the patient model before the update
+	cacheError error
```

Comment:
I think that this may cause problems.  HTTP requests are processed in a multi-threaded manner.  So it could be that the interceptor is being invoked simultaneously by multiple incoming requests.  In this case, `patientBefore` might be overwritten before the `After` function is called.

In hind sight, we may have wanted to implement some type of `Context` passing in the interceptor framework (essentially, the `Before` method is passed a `Context` object that it can set values in; then that same `Context` object is passed to the `After` function so it can pull things out of it).

For now, however, you can probably get away with storing the "beforePatient" in a map for which the patient's `Id` is the key.  Then you only run into problems if the same patient is updated exactly concurrently in multiple requests.  But... I don't recall if Go maps are thread-safe, so you'll want to look at that too.